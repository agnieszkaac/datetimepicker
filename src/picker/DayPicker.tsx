import React from "react";
import cx from "classnames";
import moment from "moment";

import { DayPickerProps, DayProps } from "../types";

const Day: React.FunctionComponent<DayProps> = ({
  value,
  selected,
  isToday,
  onPick,
}) => (
  <button
    className={cx("day", { selected: selected, "is-today": isToday })}
    onClick={onPick}
    value={value}
  >
    {value}
  </button>
);

export const DayPicker: React.FunctionComponent<DayPickerProps> = ({
  value,
  monthLength,
  onPick,
}) => (
  <div className="day-picker">
    {Array.from({ length: monthLength }, (e, i) => (
      <Day
        key={i + 1}
        value={(i + 1).toString()}
        selected={value === (i + 1).toString()}
        isToday={moment().date().toString() === value}
        onPick={onPick}
      />
    ))}
  </div>
);
