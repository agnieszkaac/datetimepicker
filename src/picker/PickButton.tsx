import React from "react";
import { Moment } from "moment";

export type PickButtonProps = {
  date: Moment;
  format: string;
  className?: string;
  onClick: (e: React.MouseEvent<HTMLButtonElement>) => void;
};

export const PickButton: React.FC<PickButtonProps> = (props) => {
  const { date, format, className, onClick } = props;
  return (
    <button
      type="button"
      value={date.toString()}
      className={className}
      onClick={onClick}
    >
      {date.format(format)}
    </button>
  );
};
