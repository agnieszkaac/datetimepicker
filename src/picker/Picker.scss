.picker {
  display: flex;
  flex-flow: column;
  justify-content: center;
  align-items: center;
  min-width: 200px;
  padding: 4px;
  margin-top: 4px;
  background-color: white;
  border: 1px solid #e2e2e2;
  border-radius: 2px;
}
