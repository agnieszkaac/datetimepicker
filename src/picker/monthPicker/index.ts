export * from "./Month";
export * from "./MonthPicker";