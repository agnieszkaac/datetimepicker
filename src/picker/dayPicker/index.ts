import { Day } from "./Day";
import DayPicker from "./DayPicker";
import { DayProps, DayPickerProps } from "./types";

export { Day, DayPicker, DayProps, DayPickerProps };
