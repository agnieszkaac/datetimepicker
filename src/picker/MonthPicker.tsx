import React from "react";
import moment from "moment";
import cx from "classnames";

import { PickerComponentProps } from "./types";
import { defaultGridSize as length } from "./utils";
import { DateContext, ViewContext } from "../state";

import { PickButton } from "./PickButton";

import "./MonthPicker.scss";

export const MonthPicker: React.FC<PickerComponentProps> = ({ onPick }) => {
  const { viewDate } = React.useContext(ViewContext);
  const { date } = React.useContext(DateContext);

  return (
    <div className="month-picker">
      {Array.from({ length }, (_e, i) => {
        const month = moment(viewDate).month(i);
        return (
          <PickButton
            key={i}
            type="month"
            date={month.startOf("month")}
            format="MMM"
            className={cx({
              selected: date?.month() === i && date?.isSame(viewDate, "y"),
              "is-today": month.isSame(moment(), "M"),
            })}
            onClick={onPick}
          />
        );
      })}
    </div>
  );
};
