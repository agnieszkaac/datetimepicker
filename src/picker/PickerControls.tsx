import React from "react";
import { Moment } from "moment";

import { PickerComponentProps } from "./types";
import { DateContext, ViewContext } from "../state";

import { PickButton } from "./PickButton";

const displayFormat = {
  day: "D",
  month: "MMM",
  year: "YYYY",
};

type PickerControlsProps = PickerComponentProps & {
  length: number;
  type: "day" | "month" | "year";
  dateFormula: (viewDate: Moment, i: number) => Moment;
  getClassName: (date: Moment, i: number) => string;
};

export const PickerControls: React.FC<PickerControlsProps> = (props) => {
  const { length, type, dateFormula, getClassName, onPick } = props;
  const { viewDate } = React.useContext(ViewContext);
  const { date } = React.useContext(DateContext);

  return (
    <>
      {Array.from({ length }, (_e, i) => {
        const dateValue = dateFormula(viewDate, i);
        return (
          <PickButton
            key={i}
            type={type}
            date={dateValue}
            format={displayFormat[type]}
            onClick={onPick}
            className={getClassName(dateValue, i)}
          />
        );
      })}
    </>
  );
};
