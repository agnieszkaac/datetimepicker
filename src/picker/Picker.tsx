import React, { Ref, useState } from "react";
import moment, { Moment } from "moment";

import { getNextView, getViewDate, View } from "./utils";
import { MonthPicker } from "./monthPicker";
import { DayPicker } from "./dayPicker";
import { YearPicker } from "./yearPicker";
import { Switch } from "./switch/Switch";
import { configLocale } from "../utils";

import "./Picker.scss";

export type PickerProps = {
  date: Moment | undefined;
  locale: string;
  firstDayOfWeek: number | undefined;
  pickerRef: Ref<HTMLDivElement>;
  onPick: (e: React.MouseEvent<HTMLButtonElement>) => void;
}

export const Picker: React.FunctionComponent<PickerProps> = ({
  date,
  locale,
  firstDayOfWeek,
  pickerRef,
  onPick,
}) => {
  const [view, setView] = useState(View.Day);
  const [viewDate, setViewDate] = useState(moment(date).startOf("M"));

  // configLocale(viewDate, locale, firstDayOfWeek);

  const handlePick = (e: React.MouseEvent<HTMLButtonElement>) => {
    if (view === View.Day) {
      onPick(e);
      return;
    }
    setViewDate(moment(e.currentTarget.value).startOf("M"));
    setView(getNextView(-1, view));
  };

  const handleRangeSwitch = (e: React.MouseEvent<HTMLButtonElement>) => {
    setViewDate(getViewDate(Number(e.currentTarget.value), view, viewDate));
  };

  const PickerComponent =
    view === View.Day
      ? DayPicker
      : view === View.Month
      ? MonthPicker
      : YearPicker;

  return (
    <div className="picker" ref={pickerRef}>
      <Switch
        view={view}
        viewDate={viewDate}
        onViewSwitch={() => setView(getNextView(1, view))}
        onRangeSwitch={handleRangeSwitch}
      />
      <PickerComponent date={date} viewDate={viewDate} onPick={handlePick} />
    </div>
  );
};
