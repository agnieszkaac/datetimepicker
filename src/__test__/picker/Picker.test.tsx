import React from "react";
import MockDate from "mockdate";
import moment from "moment";
import { render, fireEvent } from "@testing-library/react";

import { Picker, PickerProps } from "../../picker/Picker";

describe("<Picker /> should", () => {
  let props: PickerProps;

  beforeEach(() => {
    MockDate.set("2020-01-01");
    props = {
      date: moment(),
      locale: "en",
      firstDayOfWeek: 0,
      pickerRef: null,
      onPick: jest.fn(),
    };
  });

  afterEach(() => {
    MockDate.reset();
  });

  it("match snapshot", () => {
    const { asFragment } = render(<Picker {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("switch between picker views using view switcher and date buttons", () => {
    const { getByTestId, getAllByTestId } = render(<Picker {...props} />);
    const viewSwitch = getByTestId("view-switch");

    expect(getByTestId("day-picker")).toBeInTheDocument();

    //day -> month
    fireEvent.click(viewSwitch);
    expect(getByTestId("month-picker")).toBeInTheDocument();

    //month -> year
    fireEvent.click(viewSwitch);
    expect(getByTestId("year-picker")).toBeInTheDocument();

    //year -> month
    fireEvent.click(getAllByTestId("year")[0]);
    expect(getByTestId("month-picker")).toBeInTheDocument();

    //month -> day
    fireEvent.click(getAllByTestId("month")[0]);
    expect(getByTestId("day-picker")).toBeInTheDocument();
  });

  describe("switch ranges", () => {
    it("within day view", () => {
      const { getAllByTestId } = render(<Picker {...props} />);
      const rangeSwitches = getAllByTestId("range-switch");
      const days = getAllByTestId("day");

      expect(days[0]).toHaveValue("2019-12-28T23:00:00.000Z");

      fireEvent.click(rangeSwitches[0]);

      expect(days[0]).toHaveValue();

    });
    it("within month view", () => {

    });
    it("within year view", () => {

    })
  })
});
