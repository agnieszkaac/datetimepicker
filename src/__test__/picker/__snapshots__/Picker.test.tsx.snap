// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Picker /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="picker"
  >
    <div
      class="switch"
    >
      <button
        class="range-switch"
        data-testid="range-switch"
        value="-1"
      >
        <span>
          ❮
        </span>
      </button>
      <button
        class="view-switch"
        data-testid="view-switch"
      >
        January 2020
      </button>
      <button
        class="range-switch"
        data-testid="range-switch"
        value="1"
      >
        <span>
          ❯
        </span>
      </button>
    </div>
    <div
      class="week-days"
    >
      <span
        class="week-day"
        data-testid="week-day"
      >
        Sun
      </span>
      <span
        class="week-day"
        data-testid="week-day"
      >
        Mon
      </span>
      <span
        class="week-day"
        data-testid="week-day"
      >
        Tue
      </span>
      <span
        class="week-day"
        data-testid="week-day"
      >
        Wed
      </span>
      <span
        class="week-day"
        data-testid="week-day"
      >
        Thu
      </span>
      <span
        class="week-day"
        data-testid="week-day"
      >
        Fri
      </span>
      <span
        class="week-day"
        data-testid="week-day"
      >
        Sat
      </span>
    </div>
    <div
      class="day-picker"
      data-testid="day-picker"
    >
      <button
        class="day prev-month"
        data-testid="day"
        type="button"
        value="2019-12-28T23:00:00.000Z"
      >
        29
      </button>
      <button
        class="day prev-month"
        data-testid="day"
        type="button"
        value="2019-12-29T23:00:00.000Z"
      >
        30
      </button>
      <button
        class="day prev-month"
        data-testid="day"
        type="button"
        value="2019-12-30T23:00:00.000Z"
      >
        31
      </button>
      <button
        class="day selected is-today"
        data-testid="day"
        type="button"
        value="2019-12-31T23:00:00.000Z"
      >
        1
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-01T23:00:00.000Z"
      >
        2
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-02T23:00:00.000Z"
      >
        3
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-03T23:00:00.000Z"
      >
        4
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-04T23:00:00.000Z"
      >
        5
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-05T23:00:00.000Z"
      >
        6
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-06T23:00:00.000Z"
      >
        7
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-07T23:00:00.000Z"
      >
        8
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-08T23:00:00.000Z"
      >
        9
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-09T23:00:00.000Z"
      >
        10
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-10T23:00:00.000Z"
      >
        11
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-11T23:00:00.000Z"
      >
        12
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-12T23:00:00.000Z"
      >
        13
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-13T23:00:00.000Z"
      >
        14
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-14T23:00:00.000Z"
      >
        15
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-15T23:00:00.000Z"
      >
        16
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-16T23:00:00.000Z"
      >
        17
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-17T23:00:00.000Z"
      >
        18
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-18T23:00:00.000Z"
      >
        19
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-19T23:00:00.000Z"
      >
        20
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-20T23:00:00.000Z"
      >
        21
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-21T23:00:00.000Z"
      >
        22
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-22T23:00:00.000Z"
      >
        23
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-23T23:00:00.000Z"
      >
        24
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-24T23:00:00.000Z"
      >
        25
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-25T23:00:00.000Z"
      >
        26
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-26T23:00:00.000Z"
      >
        27
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-27T23:00:00.000Z"
      >
        28
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-28T23:00:00.000Z"
      >
        29
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-29T23:00:00.000Z"
      >
        30
      </button>
      <button
        class="day"
        data-testid="day"
        type="button"
        value="2020-01-30T23:00:00.000Z"
      >
        31
      </button>
      <button
        class="day next-month"
        data-testid="day"
        type="button"
        value="2020-01-31T23:00:00.000Z"
      >
        1
      </button>
      <button
        class="day next-month"
        data-testid="day"
        type="button"
        value="2020-02-01T23:00:00.000Z"
      >
        2
      </button>
      <button
        class="day next-month"
        data-testid="day"
        type="button"
        value="2020-02-02T23:00:00.000Z"
      >
        3
      </button>
      <button
        class="day next-month"
        data-testid="day"
        type="button"
        value="2020-02-03T23:00:00.000Z"
      >
        4
      </button>
      <button
        class="day next-month"
        data-testid="day"
        type="button"
        value="2020-02-04T23:00:00.000Z"
      >
        5
      </button>
      <button
        class="day next-month"
        data-testid="day"
        type="button"
        value="2020-02-05T23:00:00.000Z"
      >
        6
      </button>
      <button
        class="day next-month"
        data-testid="day"
        type="button"
        value="2020-02-06T23:00:00.000Z"
      >
        7
      </button>
      <button
        class="day next-month"
        data-testid="day"
        type="button"
        value="2020-02-07T23:00:00.000Z"
      >
        8
      </button>
    </div>
  </div>
</DocumentFragment>
`;
